sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala

```diff
@@ -773,24 +773,44 @@ case class GpuLike(left: Expression, right: Expression, escapeChar: Char)
 object GpuRegExpUtils {
 
   /**
-   * Determine if a string contains back-references such as `$1` but ignoring
-   * if preceded by escape character.
+   * Convert symbols of back-references if input string contains any.
+   * In spark's regex rule, there are two patterns of back-references:
+   * \group_index and $group_index
+   * This method transforms above two patterns into cuDF pattern ${group_index}, except they are
+   * preceded by escape character.
+   *
+   * @param rep replacement string
+   * @return A pair consists of a boolean indicating whether containing any backref and the
+   *         converted replacement.
    */
-  def containsBackrefs(s: String): Boolean = {
+  def backrefConversion(rep: String): (Boolean, String) = {
+    val b = new StringBuilder
     var i = 0
-    while (i < s.length) {
-      if (s.charAt(i) == '\\') {
+    while (i < rep.length) {
+      // match $group_index or \group_index
+      if (Seq('$', '\\').contains(rep.charAt(i))
+        && i + 1 < rep.length && rep.charAt(i + 1).isDigit) {
+
+        b.append("${")
+        var j = i + 1
+        while (j + 1 < rep.length && rep.charAt(j).isDigit) {
+          b.append(rep.charAt(j))
+          j += 1
+        }
```

Comment:
This code seems to be incorrect and does not handle the case where the pattern ends with `$1` because `j` would the index of the `1` and `j+1 < rep.length` would be false so it does not append the `1` to the cuDF pattern and produces `${}` instead.

Reply:
Yes, it's a blunder, which leads to the failure of test cases like `'REGEXP_REPLACE(a, "(TEST)", "$1")'`. 
I have fixed it.